Document the current MIR semantics that are clear from existing code

JakobDegen opened the pull request:
This PR adds documentation to places, operands, rvalues, statementkinds, and terminatorkinds that describes their existing semantics and requirements. In many places the semantics depend on the Rust memory model or other T-Lang decisions - when this is the case, it is just noted as such with links to UCG issues where possible. I'm hopeful that none of the documentation added here can be used to justify optimizations that depend on the memory model. The documentation for places and operands probably comes closest to running afoul of this - if people think that it cannot be merged as is, it can definitely also be taken out.

The goal here is to only document parts of MIR that seem to be decided already, or are at least depended on by existing code. That leaves quite a number of open questions - those are marked as "needs clarification." I'm not sure what to do with those in this PR - we obviously can't decide all these questions here. Should I just leave them in as is? Take them out? Keep them in but as `//` instead of `///` comments?

If this is too big to review at once, I can split this up.

r? rust-lang/mir-opt

JakobDegen commented:
I've left a bunch of FIXME's in here. I may try and get to some of them before merging, but they should all be non-critical and can just be added later.

Also, the test change has a [relevant Zulip conversation](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/Using.20Intrinsics.20Generically)

scottmcm commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2447,134 @@ impl<'tcx> Operand<'tcx> {
+
+    /// Exactly like `BinaryOp`, but less operands.
+    ///
+    /// Also does two's-complement arithmetic. Negation requires a signed integer or a float; binary
```
tiny nit: maybe use "bitwise" instead of "binary", since when talking about "binary/unary operations" my brain read `not(a, b)` and got confused.
```suggestion
    /// Also does two's-complement arithmetic. Negation requires a signed integer or a float; bitwise
```

JakobDegen replied:
I had written that, thought it looked wrong, but then couldn't figure it out lol. Will fix

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1762,8 +1762,98 @@ pub struct CopyNonOverlapping<'tcx> {
+/// **Needs clarification**: What about metadata resulting from dereferencing wide pointers (and
+/// possibly from accessing unsized locals - not sure how those work)? That probably deserves to go
+/// on the list above and be discussed too. It is also probably necessary for making the indexing
+/// stuff lass hand-wavey.
```
*less

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1762,8 +1762,98 @@ pub struct CopyNonOverlapping<'tcx> {
+///    pointed to type.
+///
+/// **Needs clarification**: What about metadata resulting from dereferencing wide pointers (and
+/// possibly from accessing unsized locals - not sure how those work)? That probably deserves to go
```
A place stores an address and an optional meta value.

JakobDegen replied:
Yeah, this is probably the right thing to have. That being said, I'm going to leave this open until I get a chance to more thoroughly go through all of the documentation and deal with DSTs (or probably someone else should do that because I know very little about it)

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2365,57 +2365,134 @@ impl<'tcx> Operand<'tcx> {
+/// Not all of these are allowed at every [`MirPhase`] - when this is the case, it's stated below.
+///
+/// Computing any rvalue begins by evaluating the places and operands in the rvalue in the order in
+/// which they appear. These are then used to produce a "value" - the same kind of value that an
```
I think it should be left undefined in which order theybare evaluated.

JakobDegen replied:
I've changed the ordering question to a "needs clarification" section in an effort to only document things which are already decided.

That being said, I don't think this is actually an option. First of all, I'm not sure "order is left indeterminate" implies "any attempt on depending on the order is UB" - it might be the case that we instead get two distinct well-defined possibilities. More importantly though, this doesn't feel like a thing that MIRI can implement, which is concerning.

I also can't produce an example, but I would not be surprised if this is actually required for current Rust, since the evaluation order in Rust is well-defined. I can't get MIR building to produce such MIR, but it would at least imply that the temporaries produced by this are required:
```rust
pub fn foo(a: &mut i32) -> i32 {
    let derived = &mut *a;
    *derived + *a
}
```

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2365,57 +2365,134 @@ impl<'tcx> Operand<'tcx> {
 
-    /// [x; 32]
+    /// Creates an array where each element is the value of the operand. This currently does not
+    /// drop the value even if the number of repetitions is zero, see [#74836].
```
Drops are always done using drop terminators and never by statements.

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2365,57 +2365,134 @@ impl<'tcx> Operand<'tcx> {
     Cast(CastKind, Operand<'tcx>, Ty<'tcx>),
 
+    /// * `Offset` has the same semantics as [`offset`](pointer::offset), except that the second
+    ///   paramter may be a `usize` as well.
```
*parameter

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2365,57 +2365,134 @@ impl<'tcx> Operand<'tcx> {
+    ///   raw pointers, or function pointers and return a `bool`.
+    /// * Left and right shift operations accept signed or unsigned integers not necessarily of the
+    ///   same type and return a value of the same type as their LHS. For all other operations, the
+    ///   types of the operands must match.
```
These mask the upper bits of the shift amount as necessary to avoid overflow. Eg shifting a 16bit int will result in all but the lower 4 bits being masked.

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2365,57 +2365,134 @@ impl<'tcx> Operand<'tcx> {
+    /// allowed here.
+    ///
+    /// **FIXME**: What about division/modulo? Are they allowed here at all? Are zero divisors still
+    /// UB? Also, which other combinations of types are disallowed?
```
Ony add, sub, mul and shifts are allowed.

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2365,57 +2365,134 @@ impl<'tcx> Operand<'tcx> {
+    /// UB? Also, which other combinations of types are disallowed?
     CheckedBinaryOp(BinOp, Box<(Operand<'tcx>, Operand<'tcx>)>),
 
+    /// Yields the size or alignment of the type as a `usize`.
```
I think this should defer to the NullOp variant for that. In the past there was also a NullOp for box allocation.

bjorn3 commented on compiler/rustc_const_eval/src/transform/validate.rs:
```diff
@@ -36,6 +35,13 @@ pub struct Validator {
+        // terribly important that they pass the validator. However, I think other passes might
+        // still see them, in which case they might be surprised. It would probably be better if we
+        // didn't put this through the MIR pipeline at all.
+        if matches!(body.source.instance, InstanceDef::Intrinsic(..) | InstanceDef::Virtual(..)) {
```
intrinsics may be codegened in the future.

JakobDegen replied:
Indeed. I think at that point this check should be refined to verify those bodies that do get codegened only. Even better would be if code elsewhere got changed to stop passing intrinsics that don't get codegened through the MIR validator. cc @oli-obk

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1560,34 +1560,79 @@ impl Statement<'_> {
+    /// a local, there exist at least one path from the root to that use that contains a
+    /// `StorageLive` more recently than a `StorageDead`.
+    ///
+    /// **Needs clarification**: Is it permitted to `StorageLive` a local for which we previously
```
Yes, can happen in loops.

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1560,34 +1560,79 @@ impl Statement<'_> {
+    /// executed `StorageDead`? How about two `StorageLive`s without an intervening `StorageDead`?
+    /// Two `StorageDead`s without an intervening `StorageLive`? LLVM says yes, poison, yes. If the
+    /// answer to any of these is "no," is breaking that rule UB or is it an error to have a path in
+    /// the CFG that might do this?
```
Miri should be made to match whatever is decided. It is currently very relaxed.

bjorn3 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1560,34 +1560,79 @@ impl Statement<'_> {
+    ///
+    /// **Needs clarification**: We currently require that the LHS place not overlap with any place
+    /// read as part of computation of the RHS. This requirement is under discussion in [#68364]. As
+    /// a part of this discussion, it is also unclear in what order the components are evaluated.
```
I think order should be left indetermimate so any attempt on depending on the order is UB.

JakobDegen replied:
(see [the comment above](https://github.com/rust-lang/rust/pull/95320#discussion_r835831170) - basically the same concerns apply here)

RalfJung replied:
I think leaving evaluation order indeterminate is a bad idea. We should specify some order eventually, and probably it should be left-to-right. (That's what Miri does.)

The disjointness thing can be enforced without leaving the order indeterminate, using the aliasing model instead. That's better because it can be checked by Miri.

bjorn3 commented on compiler/rustc_middle/src/mir/terminator.rs:
```diff
@@ -187,7 +248,7 @@ pub enum TerminatorKind<'tcx> {
         args: Vec<Operand<'tcx>>,
-        /// Destination for the return value. If some, the call is converging.
+        /// Destination for the return value. If none, the call necessarily diverges.
         destination: Option<(Place<'tcx>, BasicBlock)>,
```
Must not alias the place given to any Operand::Move passed as argument. All codegen backends depend on this property to allow passing Operand::Move as pointer in case of a type not fitting inside registers without having to male a copy on the stack.

JakobDegen replied:
I've added in a mention of that in the "needs clarification" section above. Based on the status of the linked issue, it seems that it is still unclear how we are justifying that optimization

bjorn3 commented on compiler/rustc_const_eval/src/transform/validate.rs:
```diff
@@ -664,18 +667,29 @@ impl<'a, 'tcx> Visitor<'tcx> for TypeChecker<'a, 'tcx> {
+            TerminatorKind::Resume | TerminatorKind::Abort => {
+                let bb = location.block;
+                if !self.body.basic_blocks()[bb].is_cleanup {
+                    self.fail(location, "Cannot `Resume` from non-cleanup basic block")
```
this says resume even for an abort terminator.

SkiFire13 commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1611,9 +1682,20 @@ pub enum StatementKind<'tcx> {
-    /// Denotes a call to the intrinsic function copy_overlapping, where `src_dst` denotes the
-    /// memory being read from and written to(one field to save memory), and size
-    /// indicates how many bytes are being copied over.
+    /// Denotes a call to the intrinsic function `copy_overlapping`.
+    ///
+    /// First, all three operands are evaluated. `src` and `dest` must each be a reference, pointer,
+    /// or `Box` pointing to the same type `T`. `count` must evaluate to a `usize`. Then, `src` and
+    /// `dest` are dereferenced, and `count * size_of::<T>()` bytes beginning with the first byte of
+    /// the `src` place are copied to the continguous range of bytes beginning with the first byte
+    /// of `dest`.
+    ///
+    /// **Needs clarification**: In what order are operands computed and dereferenced? It should
+    /// probably match the order for assignment, but that is also undecided.
+    ///
+    /// **Needs clarification**: Is this typed or not, ie is there a place to value and back
+    /// conversion involved? I vaguely remember Ralf saying somewhere that he thought it should not
+    /// be.
     CopyNonOverlapping(Box<CopyNonOverlapping<'tcx>>),
```
The documentation says `copy_overlapping`, but the variant is called `CopyNonOverlapping`

JakobDegen commented:
cc @RalfJung - want to check that everything written here (of specific concern is the place and operand stuff) is at least compatible with your model of MIR

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1560,34 +1582,78 @@ impl Statement<'_> {
+    /// Assign statements roughly correspond to an assignment in Rust proper (`x = ...`) except
+    /// without the possibility of dropping the previous value (that must be done separately, if at
+    /// all). The *exact* way this works is undecided. It probably does something like evaluating
+    /// the LHS and RHS, and then doing the inverse of a place to value conversion to write the
```
FWIW I think "place to value conversion" is a terrible term. This operation accesses memory! It shouldn't be describes as if it was a simple cast.

So I'd say something more along the lines of: it evaluates the LHS and RHS to places, loads a value from the place denoted by the RHS, and then stores that value into the place denoted by the LHS.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1560,34 +1582,78 @@ impl Statement<'_> {
+    /// **Needs clarification**: Is it permitted to have two `StorageLive`s without an intervening
+    /// `StorageDead`? Two `StorageDead`s without an intervening `StorageLive`? LLVM says poison,
+    /// yes. If the answer to any of these is "no," is breaking that rule UB or is it an error to
+    /// have a path in the CFG that might do this?
```
I don't think CFG paths are relevant. What's relevant is the actual sequence of statements that are executed when the function runs. There's basically a boolean flag dead-or-alive for each local that starts out 'dead' for locals with annotations, and that is set appropriately by these annotations. Accessing (evaluating the place expression) of a local is UB if it is currently 'dead'.

Actually it's more than that, each StorageLive allows the local to be put at a different location in memory.

JakobDegen replied:
> I don't think CFG paths are relevant

Well, because of the `MaybeStorageLiveLocals` in the validator, this is already the case. Certainly CFG paths should not be relevant in the spec/miri, but the docs here have to also mention what is required by the validator, which might have a stronger opinion

tmiasko replied:
I think the storage check currently present in the validator should be considered to be merely a debugging aid. It is possible to write unsafe code that uses local when it does not have storage. With additional optimizations present such code would fail the check. When we get to that point the check should be disabled.

JakobDegen replied:
Ah, this is a good point. With the current setup, we can't turn `*r` into `x` even if we know that `r = &mut x;`, because that might turn UB into an ICE. I'll mention this in the documentation

scottmcm replied:
Come to think of it, we should probably mention that we often don't want to "remove" UB in MIR optimizations in general, since keeping it visible to the codegen backend is important.

For example, we don't want to lower `switchInt [0 => bb3] otherwise: bbUnreachable` to `goto bb3`, since that keeps LLVM from knowing about the impossible condition.

JakobDegen replied:
Yeah, these kinds of things should definitely be documented, but imo not here. I do want to start an organized way of defining "canonicalized MIR" at some point, because I think that will be increasingly important, and this seems like the kind of thing to include there

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1762,8 +1843,98 @@ pub struct CopyNonOverlapping<'tcx> {
+///
+///  1. The part of memory that is referred to (see discussion below for details).
+///  2. The type of the place and an optional variant index. See [`PlaceTy`][tcx::PlaceTy]
+///  3. The provenance with which the place is being accessed.
```
Having 2 between 1 and 3 seems odd. 1 and 3 together make a pointer (like any pointer *value* in the program) and that is probably worth pointing out.

RalfJung replied:
Oh also, places can contain metadata if they are unsized (like pointers).

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1762,8 +1843,98 @@ pub struct CopyNonOverlapping<'tcx> {
+///    have type `usize`.
+///  - [`Deref`](ProjectionElem::Deref): Derefs are the last type of projection, and the most
+///    complicated. They are only legal on parent places that are references, pointers, or `Box`. A
+///    `Deref` projection begins by creating a value from the parent place, as if by
```
```suggestion
///    `Deref` projection begins by loading a value from the parent place, as if by
```

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1762,8 +1843,98 @@ pub struct CopyNonOverlapping<'tcx> {
+///    complicated. They are only legal on parent places that are references, pointers, or `Box`. A
+///    `Deref` projection begins by creating a value from the parent place, as if by
+///    [`Operand::Copy`]. It then dereferences the resulting pointer, creating a place of the
+///    pointed to type.
```
Note that this operation can cause UB: the crated place must be dereferencable and suitably aligned, as determined by its type.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1762,8 +1843,98 @@ pub struct CopyNonOverlapping<'tcx> {
+/// One possible model that I believe makes sense is that "part of memory" is actually just the
+/// address of the beginning of the referred to range of bytes. For sized types, the size of the
+/// range is then stored in the type, and for unsized types it's stored (possibly indirectly,
+/// through a vtable) in the metadata.
```
That's what Miri does, and I think it is the right model.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2132,24 +2303,42 @@ pub struct SourceScopeLocalData {
+///
+/// [value-def]: https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/value-domain.md
+///
+/// The most common way to create values is via a place to value conversion. A place to value
```
Ah now this is the right place for the rant that I put earlier. ;)

I don't think we should all this a "conversion". That sounds like an `as` cast, like just putting something into a different representation. That's not at all what happens though. This is a *load* from memory, and should be described as that.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2132,24 +2303,42 @@ pub struct SourceScopeLocalData {
+/// This is what is implemented in miri today. Are these the semantics we want for MIR? Is this
+/// something we can even decide without knowing more about Rust's memory model?
+///
+/// A place to value conversion on a place that has its variant index set is not well-formed.
```
Why is this needed? Miri at least is fine with such places being loaded from. The result is basically the same as loading without the variant index set (but the variant index is 'carried over' to the type information of the created value).

JakobDegen replied:
I vaguely remember seeing some code that would ICE if this was the case, but now can't find it. In any case, I can't think of a good reason to do this, and being able to check this in the validator seems like it could catch some bugs. I can change this to a "needs clarification" if this is controversial though.

RalfJung replied:
It seems a bit too arbitrary to be canonical, so yeah let's move it to "needs clarification".

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2446,134 @@ impl<'tcx> Operand<'tcx> {
+///
+/// Computing any rvalue begins by evaluating the places and operands in some order (**Needs
+/// clarification**: Which order?). These are then used to produce a "value" - the same kind of
+/// value that an [`Operand`] is.
```
```suggestion
/// value that an [`Operand`] evaluates to.
```
or "produces" or so?

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2446,134 @@ impl<'tcx> Operand<'tcx> {
+    /// `*const T`, and if neither of those apply a `&T`.
+    ///
+    /// **Note:** This is a runtime operation that actually executes code and is in this sense more
+    /// like a function call. Also, DSEing these causes `fn main() {}` to SIGILL for some reason
```
"DSE" = Dead Store Elimination? Might be worth spelling out to make this a bit more accessible.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2446,134 @@ impl<'tcx> Operand<'tcx> {
+    ///
+    /// **Note:** This is a runtime operation that actually executes code and is in this sense more
+    /// like a function call. Also, DSEing these causes `fn main() {}` to SIGILL for some reason
+    /// that I never got a chance to look into.
```
```suggestion
    /// that I (JakobDegen) never got a chance to look into.
```

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2446,134 @@ impl<'tcx> Operand<'tcx> {
+    ///   raw pointers, or function pointers and return a `bool`.
+    /// * Left and right shift operations accept signed or unsigned integers not necessarily of the
+    ///   same type and return a value of the same type as their LHS. For all other operations, the
+    ///   types of the operands must match. Like in Rust, the RHS is truncated as needed.
```
The last sentence refers to shifts only, right? It should be moved up. Also please make shift their own bullet.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2446,134 @@ impl<'tcx> Operand<'tcx> {
+
+    /// Same as `BinaryOp`, but yields `(T, bool)` instead of `T`. In addition to performing the
+    /// same computation as the matching `BinaryOp`, checks if the infinite precison result would be
+    /// unequal to the actual result and sets the `bool` if this is the case.
```
That's not entirely true -- floating point operations always return `false` even if there were rounding errors (which means the "infinite precision result" does not match what has been returned).

JakobDegen replied:
I just double checked, they're not supported here

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2257,57 +2446,134 @@ impl<'tcx> Operand<'tcx> {
     ///
-    /// This is different a normal transmute because dataflow analysis will treat the box
-    /// as initialized but its content as uninitialized.
+    /// This is different a normal transmute because dataflow analysis will treat the box as
```
```suggestion
    /// This is different from a normal transmute because dataflow analysis will treat the box as
```

RalfJung commented:
> cc @RalfJung - want to check that everything written here (of specific concern is the place and operand stuff) is at least compatible with your model of MIR

I left a bunch of comments. Overall, sounds good!
I didn't look at the validator changes.

tmiasko commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1560,34 +1582,78 @@ impl Statement<'_> {
+    /// could meaningfully require this anyway. How about free lifetimes? Is ignoring this
+    /// interesting for optimizations? Do we want to allow such optimizations?
+    ///
+    /// **Needs clarification**: We currently require that the LHS place not overlap with any place
+    /// read as part of computation of the RHS. This requirement is under discussion in [#68364]. As
```
Note that the discussion makes a distinction between primitives and non-primitives, which is important given that MIR constructed for compound assignment would already violate the non-overlap rule.

JakobDegen replied:
Can you give an example of this? I didn't find this in the discussion in the issue

tmiasko replied:
Without overflow checks:

```rust
pub fn f(mut a: u32) {
    a += 1;
}
```

```rust
// MIR for `f` 0 mir_map

fn f(_1: u32) -> () {
    debug a => _1;                       // in scope 0 at a.rs:1:10: 1:15
    let mut _0: ();                      // return place in scope 0 at a.rs:1:22: 1:22

    bb0: {
        _1 = Add(_1, const 1_u32);       // scope 0 at a.rs:2:5: 2:11
        _0 = const ();                   // scope 0 at a.rs:1:22: 3:2
        return;                          // scope 0 at a.rs:3:2: 3:2
    }
}
```

RalfJung replied:
Good point. I don't even know what "those not producing primitives" means.

But having such an already subtle rule only apply sometimes seems terrible. We should either make it apply always, or never, or have a `bool` annotated at the statement that says whether they may overlap.

tmiasko replied:
On the one hand MIR building constructs assignments where lhs and rhs are overlapping. On the other hand the LLVM code generation for an assignment uses `llvm.memcpy` which requires destination to be either equal or non-overlapping with source.

RalfJung replied:
Well yeah that's bad and fragile and we shouldn't do it IMO. ;)

JakobDegen commented:
Added an extra commit with a bunch of changes in response to comments and rebased. The biggest things that changed are
 1. "Place to value" conversion got renamed to "loading a place" or something similar. Besides being clearer, its also easier to fit into the flow of a sentence, so that's a plus.
 2. I removed a "needs clarification" section and instead put in that places contain addresses to the start of the place specifically. I can revert this if it turns out to be controversial and/or too much as part of this big a PR.
 3. I added metadata in the place documentation, and descriptions for how its computed. I did not try documenting how unsized locals work, mostly because I don't understand it, and just left a note in.

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1858,41 +1858,55 @@ pub struct CopyNonOverlapping<'tcx> {
+/// Furthermore, MIR produces assignments to unsized locals, although that is not permitted under
+/// `#![feature(unsized_locals)]` in Rust. Besides just putting "unsized locals are special and
+/// different" in a bunch of places, I (JakobDegen) don't know how to incorporate this behavior into
+/// the current MIR semantics in a clean way - possibly this needs some design work first.
```
Yeah. We have a hack in Miri that I can explain to you, but I would love for someone to come up with a cleaner solution. :)

RalfJung commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1775,8 +1856,82 @@ pub struct CopyNonOverlapping<'tcx> {
+///    pointee type is unsized, the pointer additionally stored the value of the metadata.
+///
+/// Computing a place may cause UB. One possibility is that the pointer used for a `Deref` may not
+/// be suitably aligned. Another possibility is that the place is not in bouns, meaning it does not
```
```suggestion
/// be suitably aligned. Another possibility is that the place is not in bounds, meaning it does not
```

JakobDegen commented:
I've done my best to address all the outstanding concerns. One thing I had mentioned earlier that I would explicitly like to call attention to again is this:

> I'm hopeful that none of the documentation added here can be used to justify optimizations that depend on the memory model

I would like it to be the case that people can actually use this documentation to justify their opts - that means its exceedingly important that this does not say too much. I'd like to ask whoever does the final review of this to pay particular attention to that point

bors commented:
:pushpin: Commit 9745a172a9954734367c166f07bcb5a83a07878f has been approved by `oli-obk`

<!-- @bors r=oli-obk 9745a172a9954734367c166f07bcb5a83a07878f -->
<!-- homu: {"type":"Approved","sha":"9745a172a9954734367c166f07bcb5a83a07878f","approver":"oli-obk"} -->

wesleywiser commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -1775,8 +1856,82 @@ pub struct CopyNonOverlapping<'tcx> {
+///    given one, and makes no other changes. A `Downcast` projection on a place with its variant
+///    index already set is not well-formed.
+///  - [`Field`](ProjectionElem::Field): `Field` projections take their parent place and create a
+///    place referring to one of the fields of the type. The resulting address is the paren't
```
```suggestion
///    place referring to one of the fields of the type. The resulting address is the parent
```

wesleywiser reviewed (approved):
Looks good to me! 💯

Dylan-DPC commented:
@bors r- (merge conflict)

bors commented:
:umbrella: The latest upstream changes (presumably #95125) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

oli-obk commented:
@bors r+

bors commented:
:pushpin: Commit 8732bf5db372e3b9297e854ce71851bbc6a90893 has been approved by `oli-obk`

<!-- @bors r=oli-obk 8732bf5db372e3b9297e854ce71851bbc6a90893 -->
<!-- homu: {"type":"Approved","sha":"8732bf5db372e3b9297e854ce71851bbc6a90893","approver":"oli-obk"} -->
